### PR TITLE
Assignment 3 - Ian Kaiserman

### PR DESCRIPTION
**Pull request by Afterlyte**
Paired with @eroadhouse

EC attempt on Write unit test

### CODE REVIEW

**Comment by zlawson-ewu** [comment]
I will review this.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 1]
```diff
@@ -1,10 +1,61 @@
...
+    }
+
+    [TestMethod]
+    public void ConsoleOutput_Write_WritesLine()
+    {
+        // Arrange
+        string text = "test string";
+        ConsoleOutput consoleOutput = new();
+        StringWriter writer = new();
+        Console.SetOut(writer);
```
Comment (zlawson-ewu) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:52]: I like your method of testing this. I wish I'd known about StringWriter before I thought to use a file.

**Review by zlawson-ewu — COMMENTED** [review]
Assignment:

  Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
  
  The JokeService will need to have the interface applied to it. ✔
  
  Create an implementation of the output interface that writes out the joke to the console. ✔
  
  Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
  
  The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
  
  Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌
  
  Nullability for all projects ✔
  
  Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
  
  Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
  
  For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
  
  Ensure there are no errors or warnings (including code analysis warnings) ✔
  
  All of the above should be unit tested ✔
  
  Choose simplicity over complexity ✔

Extra Credit:

  Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)? ✔
  
  The Geek jokes API that is being used can also return jokes in a JSON format. Update the JokeService to retrieve jokes using JSON. GetJoke should still return a string. ❌

Comments:
  Overall good work! All tests passed, no warnings. Running Main() works as intended. The ConsoleOutput class is tested in a nice and simple way. The code coverage is 100% for Jester and ConsoleOutput, though I don't think the jester tests are quite right for reasons mentioned in the line comment. If the Verify() method is used in the tests instead, I'd consider that bullet point met.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        }
+    }
+
+    public string TellJoke()
+    {
+        string joke = Joke.GetJoke();
+        while (joke.Contains("Chuck Norris"))
+        {
+            joke = Joke.GetJoke();
+        }
```
[the remark this answers is not in the document]
Comment (Afterlyte) [reply CanHazFunny/CanHazFunny/Jester.cs:52]: This is a good idea, but when I went to implement it I noticed I still need the extra line to create the variable, otherwise the string obtained in the do/while loop isn't accessible to the rest of the method.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs (hunk 1 above)
Comment (Afterlyte) [reply CanHazFunny/CanHazFunny.Tests/JesterTests.cs:52]: Thank you! I remember doing something similar in Java so I did a little researched and found something similar can be done here.

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 3]
```diff
@@ -1,10 +1,61 @@
...
+            .Returns("Chuck Norris")
+            .Returns("Test joke");
+        Jester testJester = new(jokeServiceMock.Object, new ConsoleOutput());
+
+        // Act
+        string joke = testJester.TellJoke();
+
+        // Assert
+        Assert.AreEqual<string>("Test joke", joke);
+    }
```
[the remark this answers is not in the document]
Comment (Afterlyte) [reply CanHazFunny/CanHazFunny.Tests/JesterTests.cs:43]: Interesting. we'll look into it a bit, but I'm not sure how necessary it is, specifically since we also did the extra credit block of unit testing the write method on its own. Because realistically if the string is being returned it means it has to have gotten through that line of code successfully, otherwise it would throw an exception of sorts, something else would fail, etc. The main purpose of having TellJoke return a string was for testing purposes, it's the most straightforward way we could think of being able to test input/output using Moq in this case. 

Long story short I'm pretty confident in the fact that it's marked as 100% code coverage and doesn't count it as "partial coverage" or anything like that, but we'll look into getting Verifiable working

File: CanHazFunny/CanHazFunny/CanHazFunny.csproj [hunk 4]
```diff
@@ -1,8 +1,5 @@
 <Project Sdk="Microsoft.NET.Sdk">
-
-  <PropertyGroup>
-    <OutputType>Exe</OutputType>
-    <TargetFramework>net6.0</TargetFramework>
-  </PropertyGroup>
-
+  <ItemGroup>
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/CanHazFunny.csproj:2]: Next time remember lang version, target framework, nullability, NET analyzers, and treat warnings as errors.

File: CanHazFunny/CanHazFunny/IOutput.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CanHazFunny;
+
+public interface IOutput
+{
+    void Write(String text);
```
Comment (JLee35) [inline CanHazFunny/CanHazFunny/IOutput.cs:11]:
```suggestion
    void Write(string text);
```

File: CanHazFunny/Directory.Build.props [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+	<PropertyGroup>
+		<OutputType>Exe</OutputType>
+		<TargetFramework>net7.0</TargetFramework>
+		<LangVersion>11.0</LangVersion>
+		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+		<WarningsAsErrors />
+		<EnableNETAnalyzers>true</EnableNETAnalyzers>
+		<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+		<Nullable>enable</Nullable>
+	</PropertyGroup>
```
Comment (JLee35) [inline CanHazFunny/Directory.Build.props:12]: Nice!

**Comment by JLee35** [comment]
Great job!
